File: 1. 자료구조/정리/1. 순환, 반복/0. 팩토리얼/팩토리얼.c
//
//  0. 팩토리얼.c
//  Algorithm
//
//  Created by 김동현 on 7/28/25.
//

#include <stdio.h>

int factorial(int n) {
    if (n <= 1) return 1;
    return n * factorial(n-1);
}

int factorial2(int n) {
    int result = 1;
    
    for (int i=1; i<=n; i++) {
        result *= i;
    }
    
    return result;
}

int main(void) {
    int result = factorial(3);
    printf("%d\n", result);
    
    int result2 = factorial2(3);
    printf("%d\n", result2);
}

File: 1. 자료구조/정리/1. 순환, 반복/0. 팩토리얼/팩토리얼.swift
//
//  0.팩토리얼.swift
//  Algorithm
//
//  Created by 김동현 on 7/28/25.
//

import Foundation

func factorial(_ n: Int) -> Int {
    if n <= 1 { return 1 }
    return n * factorial(n - 1)
}

func factorial2(_ n: Int) -> Int {
    return (1...n).reduce(1, *)
    // return (1...n).reduce(1) { $0 * $1 }
}

@main
struct Main {
    static func main() {
        let result = factorial(3)
        print(result)
        
//        let result2 = factorial2(10)
//        print(result2)
    }
}
